src/lib/Embed.hs

```diff
@@ -4,238 +4,241 @@
 -- license that can be found in the LICENSE file or at
 -- https://developers.google.com/open-source/licenses/bsd
 
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
-
-module Embed (emit, emitTo, buildLamExpr, buildLam, buildTLam,
-              EmbedT, Embed, EmbedEnv, buildScoped, wrapDecls, runEmbedT,
-              runEmbed, zeroAt, addAt, sumAt, embedScope, withIndexed,
-              nRecGet, nTabGet, add, mul, sub, neg, div', andE,
-              select, selectAt, freshVar, unitBinder, unpackRec,
-              substEmbed, fromPair, buildLamExprAux,
-              emitExpr, unzipTab, buildFor, isSingletonType,
-              singletonTypeVal, mapScalars,
-              boolToInt, intToReal, boolToReal) where
+{-# LANGUAGE UndecidableInstances #-}
+
+module Embed (emit, emitOp, buildDepEffLam, buildLamAux, buildAbs,
+              getAllowedEffects, withEffects, modifyAllowedEffects,
+              buildLam, EmbedT, Embed, MonadEmbed, buildScoped, wrapDecls, runEmbedT,
+              runEmbed, zeroAt, addAt, sumAt, getScope, reduceBlock, withBinder,
+              app, add, mul, sub, neg, div', andE, reduceScoped, declAsScope,
+              select, selectAt, substEmbed, fromPair, getFst, getSnd,
+              emitBlock, unzipTab, buildFor, isSingletonType, emitDecl, withNameHint,
+              singletonTypeVal, mapScalars, scopedDecls, embedScoped, extendScope,
+              embedExtend, boolToInt, intToReal, boolToReal, reduceAtom,
+              unpackConsList, emitRunWriter, tabGet) where
 
 import Control.Monad
+import Control.Monad.Fail
+import Control.Monad.Except hiding (Except)
 import Control.Monad.Reader
+import Control.Monad.Writer
+import Control.Monad.Identity
+import Data.Foldable (toList)
+import Data.Maybe
 
 import Env
 import Syntax
 import Cat
 import Type
-import Subst
-import Record
 import PPrint
 
-type EmbedT m = CatT EmbedEnv m  -- TODO: consider a full newtype wrapper
-type Embed = Cat EmbedEnv
+newtype EmbedT m a = EmbedT (ReaderT EmbedEnvR (CatT EmbedEnvC m) a)
+  deriving (Functor, Applicative, Monad, MonadIO, MonadFail)
+
+type Embed = EmbedT Identity
+type EmbedEnv = (EmbedEnvR, EmbedEnvC)
+type EmbedEnvC = (Scope, [Decl])
+type EmbedEnvR = (Name, EffectRow)
```

Comment:
Can you please add some comments explaining what those things are (`EmbedEnvR` in particular)?

Reply:
Good idea.